domain/machine/state/state.go

```diff
@@ -219,3 +219,77 @@ func (st *State) AllMachineNames(ctx context.Context) ([]machine.Name, error) {
 
 	return machineNames, nil
 }
+
+// RequireMachineReboot sets the reboot flag for the specified machine, which indicates that the
+// machine requires a reboot.
+//
+// Reboot flags are handled through the "machine_requires_reboot" table which contains only
+// machine UUID for which a reboot has been requested.
+func (st *State) RequireMachineReboot(ctx context.Context, uuid string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	setRebootFlag := `INSERT INTO machine_requires_reboot (machine_uuid) VALUES ($M.machine_uuid) ON CONFLICT DO NOTHING `
+	setRebootFlagStmt, err := sqlair.Prepare(setRebootFlag, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return tx.Query(ctx, setRebootFlagStmt, sqlair.M{
+			"machine_uuid": uuid,
+		}).Run()
+	})
+	return errors.Annotatef(err, "requiring reboot of machine %q", uuid)
+}
+
+// CancelMachineReboot cancels the reboot of the specified machine by removing the corresponding flag in the database.
+func (st *State) CancelMachineReboot(ctx context.Context, uuid string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	unsetRebootFlag := `DELETE FROM machine_requires_reboot WHERE machine_uuid = $M.machine_uuid`
+	unsetRebootFlagStmt, err := sqlair.Prepare(unsetRebootFlag, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return tx.Query(ctx, unsetRebootFlagStmt, sqlair.M{
+			"machine_uuid": uuid,
+		}).Run()
+	})
+	return errors.Annotatef(err, "cancelling reboot of machine %q", uuid)
+}
+
+// IsMachineRebootRequired checks if the specified machine requires a reboot.
+//
+// It queries the "machine_requires_reboot" table for the machine UUID to determine if a reboot is required.
+// Returns a boolean value indicating if a reboot is required, and an error if any occur during the process.
+func (st *State) IsMachineRebootRequired(ctx context.Context, uuid string) (bool, error) {
+	db, err := st.DB()
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	var isRebootRequired bool
+	isRebootFlag := `SELECT true as &M.* FROM machine_requires_reboot WHERE machine_uuid = $M.machine_uuid`
```

Comment:
Just a thought, maybe `SELECT COUNT(*) ...` is better than serializing that `true`?

Reply:
I don't think it matter, but i have changed the code a bit with the introduction of `machineReboot` struct.